File: backend/internal/store/store.go
package store

import (
	"context"
	"crypto/rand"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

type Store interface {
	Health(ctx context.Context) error
	UpsertOAuthUser(ctx context.Context, user OAuthUser) (User, error)
}

type PgxStore struct {
	pool *pgxpool.Pool
}

func NewPgxStore(pool *pgxpool.Pool) *PgxStore {
	return &PgxStore{pool: pool}
}

func (s *PgxStore) Health(ctx context.Context) error {
	if s.pool == nil {
		return nil
	}
	return s.pool.Ping(ctx)
}

func (s *PgxStore) UpsertOAuthUser(ctx context.Context, user OAuthUser) (User, error) {
	_ = ctx
	// TODO: Persist users in the database (users table keyed by provider + provider_user_id).
	// TODO: Return the existing user on conflict and update profile fields.
	return newUserFromOAuth(user), nil
}

type NoopStore struct{}

func NewNoopStore() *NoopStore {
	return &NoopStore{}
}

func (s *NoopStore) Health(ctx context.Context) error {
	return nil
}

func (s *NoopStore) UpsertOAuthUser(ctx context.Context, user OAuthUser) (User, error) {
	_ = ctx
	return newUserFromOAuth(user), nil
}

type OAuthUser struct {
	Provider       string
	ProviderUserID string
	Name           string
	Email          string
	Nickname       string
	AvatarURL      string
}

type User struct {
	ID             string
	Provider       string
	ProviderUserID string
	Name           string
	Email          string
	Nickname       string
	AvatarURL      string
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

func newUserFromOAuth(oauth OAuthUser) User {
	now := time.Now().UTC()
	return User{
		ID:             userIDFromOAuth(oauth),
		Provider:       oauth.Provider,
		ProviderUserID: oauth.ProviderUserID,
		Name:           oauth.Name,
		Email:          oauth.Email,
		Nickname:       oauth.Nickname,
		AvatarURL:      oauth.AvatarURL,
		CreatedAt:      now,
		UpdatedAt:      now,
	}
}

func userIDFromOAuth(oauth OAuthUser) string {
	if oauth.Provider != "" && oauth.ProviderUserID != "" {
		return fmt.Sprintf("%s:%s", oauth.Provider, oauth.ProviderUserID)
	}
	buf := make([]byte, 8)
	if _, err := rand.Read(buf); err != nil {
		return "user_unknown"
	}
	return fmt.Sprintf("user_%x", buf)
}

File: backend/app/services/store.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone


@dataclass
class OAuthUser:
    provider: str
    provider_user_id: str
    name: str
    email: str
    nickname: str
    avatar_url: str


@dataclass
class User:
    id: str
    provider: str
    provider_user_id: str
    name: str
    email: str
    nickname: str
    avatar_url: str
    created_at: datetime
    updated_at: datetime


class Store:
    def upsert_oauth_user(self, oauth_user: OAuthUser) -> User:
        now = datetime.now(timezone.utc)
        user_id = f"{oauth_user.provider}:{oauth_user.provider_user_id}" if oauth_user.provider and oauth_user.provider_user_id else "user_unknown"
        return User(
            id=user_id,
            provider=oauth_user.provider,
            provider_user_id=oauth_user.provider_user_id,
            name=oauth_user.name,
            email=oauth_user.email,
            nickname=oauth_user.nickname,
            avatar_url=oauth_user.avatar_url,
            created_at=now,
            updated_at=now,
        )
